Greenbids RTD Module (#5613)
prebid/prebid.github.io: prebid-server/pbs-modules/greenbids-real-time-data.md

---
layout: page_v2
page_type: pbs-module
title: Prebid Server Greenbids Real Time Data Module
display_name : Greenbids Real Time Data Module
sidebarType : 5
---

# Greenbids Real Time Data Module

## Overview

Greenbids Real Time Data module filters bidders 
SSPs listed in the `imp[].ext.prebid.bidder`

> of the bid request. To perform the filtering the module uses the ML pipeline that outputs the probability 

Reviewer: ```suggestion
of the bid request. To perform the filtering the module uses AI to predict the bidding probability 
```

of bid per SSP for each `imp` for the given bid request. 

> Then this probability of bid is compared with the threshold to ensure the necessary level of filtering for a partner publisher.

Reviewer: I don't know if there is some linting rules for markdown enforced of this repository but here you have an inconsistent line wrapping.
As a personal preference, I break the paragraphs on phrases, it makes more sense during review (the line is changed if you changed the phrase, not only if you added a word two phrase earlier which require to reformat your whole paragraph).

> Then this probability of bid is compared with the threshold to ensure the necessary level of filtering for a partner publisher.

Reviewer: ```suggestion
This bidding probability is used to choose to send the bid request or not, according to the performance constraints applied by the partner publisher.
```

The RTD module uses 2 artefacts that are fetched from the Greenbids Google Cloud Storage bucket

> - ML predictor in `.onnx` format

Reviewer: ```suggestion
- ML predictor in `.onnx` format: used to predict the bidding probability
```

> - Probability thresholds in `.json` format with the list of thresholds and their corresponding target metrics

Reviewer: ```suggestion
- Probability thresholds in `.json` format: used to convert the bidding probability to a binary choice
```

### Account-Level Config

Here's an example of the account config used in PBS-Java:

```yaml
hooks:
  greenbids-real-time-data:
    enabled: true
  host-execution-plan: >
    {
      "endpoints": {
        "/openrtb2/auction": {
          "stages": {
            "processed-auction-request": {
              "groups": [
                {
                  "timeout": 100,
                  "hook-sequence": [
                    {
                      "module-code": "greenbids-real-time-data",
                      "hook-impl-code": "greenbids-real-time-data-processed-auction-request-hook"
                    }
                  ]
                }
              ]
            }
          }
        }
      }
    }
```

### List of module configuration options

- `google-cloud-greenbids-project`: Google Cloud project associated with Greenbids
- `gcs-bucket-name`: Google Cloud Storage (GCS) bucket used to fetch the artefacts ([ONNX](https://onnx.ai/) model and thresholds `.json`) necessary for prediction
- `cache-expiration-minutes`: The duration (in minutes) after which cached model and thresholds should be considered expired and refreshed

> - `geo-lite-country-path`: Geolocation capabilities to determine the country from an IP address

Reviewer: ```suggestion
- `geo-lite-country-path`: URL to the geo-ip database
```

  onnx-model-cache-key-prefix: "onnxModelRunner_"
  thresholds-cache-key-prefix: "throttlingThresholds_"
```

### Publisher bid request settings

The activated partner publishers add their configs direclty into `bid-request.json` 
where they indicate the activation of our module 
in bid request extenstion `bid-request.ext.prebid.analytics.greenbids` 
for [Analytics Reporter](https://docs.prebid.org/prebid-server/pbs-modules/greenbids-analytics-reporter.html) and 
`bid-request.ext.prebid.analytics.greenbids-rtd` for Greenbids RTD Module.

The list of the parameters necessary for RTD module activation is as follows:

| Parameter       | Scope             | Description                                                                                         | Example               | Type         |
|-----------------|-------------------|-----------------------------------------------------------------------------------------------------|-----------------------|--------------|
| pbuid           | required          | The Greenbids Publisher ID                                                                          | greenbids-publisher-1 | string       |

> | targetTpr       | required          | Ratio of passing the valid bids [0-1]                                                               | 0.9                   | float        |

Reviewer: ```suggestion
| targetTpr       | required          | Expected retained opportunities ratio [0-1]                                                            | 0.9                   | float        |
```

| explorationRate | required          | Ratio of traffic without filtering used for training ML model [0-1] (a value of 0.1 will filter 90% of the traffic) | 0.1                   | float        |

> Here's an example of how a PBS partner publisher setup using both Greenbids RTD Module and Greenbids AnalyticsReporter shoudl look like:

Reviewer: ```suggestion
Here's an example of how a PBS partner publisher setup using both Greenbids RTD Module and Greenbids AnalyticsReporter should look like:
```

```json
"ext": {
    "prebid": {
      "analytics": {
        // extension for Greenbids Analytics Reporter
        "greenbids": {
          "pbuid": "PBUID_FROM_GREENBIDS",

>           "greenbidsSampling": 1

Reviewer: The value of `1` is a very bad value (it puts our infra on load while don't bringing any value, sending a useless request for each ad call).
```suggestion
          "greenbidsSampling": 0.01
```

        },
        // extension for Greenbids Real Time Data Module
        "greenbids-rtd": {
          "pbuid": "PBUID_FROM_GREENBIDS", 
          "targetTpr": 0.95,
          "explorationRate": 0.001
        }
      }
    }
  }
```

### Enable for Spring Boot

In order to allow the module to be picked up by PBS-Java, a Spring Boot configuration property `hooks.greenbids-real-time-data.enabled` must be set to `true`.

Here's an example of how your PBS configuration YAML should look like:

```YAML
hooks:
  greenbids-real-time-data:
    enabled: true
```

## Analytics Tags

The RTD module also communicates the filtering results with the `GreenbidsAnalyticsReporter` via [AnalyticsTags](https://docs.prebid.org/prebid-server/developers/module-atags.html). 

> Here we populate analytics result of analytics tags for each `imp` the with

Reviewer: ```suggestion
Here we populate analytics result of analytics tags for each `imp` with:
```

- `fingerprint`: greenbidsId

> - `isKeptInAuction` map of booleans for each bidder whether we keep them in auction or not for the given imp

Reviewer: ```suggestion
- `isKeptInAuction`: map of booleans for each bidder whether we keep them in auction or not for the given imp
```

> - `isExploration`flag that is necessary to isolate the training data

Reviewer: ```suggestion
- `isExploration`: flag that is necessary to isolate the training data
```
